### ISSUE TITLE
Numerical error in beta_binom_pmf for high means and low fluctuations

### ISSUE THREAD

**Issue by JelleAalbers** [FlamTeam/flamedisx#37]
Flamedisx currently overestimates the differential rate at low-energy NR events under the ER model. From adapting the models_check notebook on my laptop, fixing the position at the center of the TPC, and using a 30 GeV WIMPsource, I get:

![nrdata_ermodel](https://user-images.githubusercontent.com/4354311/67161479-3f3c4a00-f35b-11e9-985d-246ab4ad7bc1.png)

This has to be the reason why, in our discrimination efficiency check, flamedisx is under some conditions outperformed by a classic (cS1, cS2) likelihood. 

Since flamedisx *over*estimates the rate, the problem can't be in the bounds calculation, so increasing max_sigma does not help.

All the other combinations (ER data + ER model, ER data + NR model, NR data + NR model, ) give consistent results:

![erdata_ermodel](https://user-images.githubusercontent.com/4354311/67161486-47948500-f35b-11e9-99fc-ec0bd63eb7a0.png)
![erdata_nrmodel](https://user-images.githubusercontent.com/4354311/67161487-47948500-f35b-11e9-82a7-69d3e4bf9718.png)
![nrdata_nrmodel](https://user-images.githubusercontent.com/4354311/67161488-49f6df00-f35b-11e9-85ad-57673e53b54d.png)

**Comment by JelleAalbers** [comment]
The problem is due to numerical errors / instabilities in fd.beta_binom_pmf. When the fluctuation amplitude becomes small, especially when the mean probability goes to 1, we get an incorrect PMF. Here is the case for mean = 97% and fluctuation = 1%. If you make the values more extreme you can get more crazy distributions.

```python
import matplotlib.pyplot as plt
import numpy as np
from multihist import Hist1d

from scipy import stats
import flamedisx as fd

# These might be introduced by the next PR
fd.MIN_FLUCTUATION_P = 0.
fd.MAX_MEAN_P = 1.

mean, sigma, n  = 0.97, 0.01, 20
n_trials = int(1e6)

a, b = fd.beta_params(mean, sigma)
p_means = stats.beta(a, b).rvs(n_trials)
ks = stats.binom(n=n, p=p_means).rvs(n_trials)
bins = np.arange(-2, 1.5 * n + 2) - 0.5
h = Hist1d(ks, bins) / n_trials
h.plot()

plt.scatter(h.bin_centers, 
            fd.beta_binom_pmf(fd.np_to_tf(h.bin_centers), n=n, p_mean=mean, p_sigma=sigma).numpy())
```
![oops](https://user-images.githubusercontent.com/4354311/67162839-72d1a100-f368-11e9-91c3-2a2b028924af.png)

The problem appears to be that the beta parameters a and b of the beta function get large, so the lgamma(a) etc. become large too, but the different lgamma terms don't differ so much in this regime.

As a workaround, I'm going to put the minimum p_electron_fluct to 0.005 and the maximum p_electron_mean to 0.95. Clearly this is a kludge, we should also support models with a very low light yield and low fluctuation amplitude. But it does fix the problem:

![nrdata_ermodel](https://user-images.githubusercontent.com/4354311/67162895-26d32c00-f369-11e9-8053-4198e92f4151.png)

Perhaps one of the earlier implementations of the beta binomial which we discarded for speed would work? I tried setting flamedisx's float type to float64, but that did not resolve the issue.

**Comment by pelssers** [comment]
As pointed out in #38, a permanent solution might be to model `p_photon_fluct = 1 - p_electron_fluct` instead of `p_electron_fluct`.
It should get rid of `MAX_MEAN_P`, but might still require `MIN_FLUCTUATION_P`

**Comment by pelssers** [comment]
Looking into this a bit further, here is a comparison of three implementations of the beta-binomial distribution using the script by @JelleAalbers. The 'true' histogram and blue 'Gamma' dots are the same as before illustrating the numerical error for `p_mean=0.97`. Added are 'Dirichlet' which implements the beta-binomial as a Dirichlet multinomial distribution with 2 classes. Lastly '1 - p_mean' computes the probability using the current 'Gamma' implementation but using the number of photons instead of electrons and `1 - p_mean` instead of `p_mean`.

![beta_binom_implementations](https://user-images.githubusercontent.com/8984694/67301584-e651ea80-f4ef-11e9-98a8-0e8016e7d0e7.png)

As can be seen both 'Dirichlet' and '1-p_mean' have a good performance for large 'p_mean' although some error remains (at index 19 none of the methods agree, but Dirichlet seems closest).

Similar results are seen with `p_mean=0.97, p_sigma=0.01`, `p_mean=0.99, p_sigma=0.01` `p_mean=0.99, p_sigma=0.005` and `p_mean=0.99, p_sigma=0.001`.

The 'Dirichlet' implementation was originally in flamedisx but replaced by the 'Gamma' method which is much faster and doesn't require explicit float64 casts.
Using '1 - p_mean' seems to be a good solution as long as 'p_mean' doesn't get small. Interestingly I don't see the method break down for small `p_electron_fluctuation` values as was reported earlier.

**Comment by JelleAalbers** [comment]
This has been addressed by #39 and #83. The ranges of allowed parameters should cover physically interesting cases. If somehow a need for ultra-low fluctuations comes up in the future, we could consider the Dirichlet implementation Bart mentioned above, though it has drawbacks as discussed.
